src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBioMartRetriever.java

```diff
@@ -0,0 +1,247 @@
+package org.reactome.addlinks.dataretrieval.ensembl;
+
+import org.apache.http.HttpException;
+import org.apache.http.HttpStatus;
+import org.json.simple.parser.ParseException;
+import org.reactome.addlinks.EnsemblBioMartUtil;
+import org.reactome.addlinks.dataretrieval.FileRetriever;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.time.Duration;
+import java.util.*;
+
+public class EnsemblBioMartRetriever extends FileRetriever {
+
+    private Properties properties;
+
+    public EnsemblBioMartRetriever() { }
+
+    public EnsemblBioMartRetriever(String retrieverName)
+    {
+        super(retrieverName);
+    }
+
+    private static final String BIOMART_SPECIES_NAME_PLACEHOLDER = "BIOMART_SPECIES_NAME";
+    private static final String BIOMART_QUERY_ID_PLACEHOLDER = "BIOMART_QUERY_ID";
+    private static final String microarrayTypesBaseQuery = String.join("&",
+            "type=listAttributes",
+            "mart=ENSEMBL_MART_ENSEMBL",
+            "virtualSchema=default",
+            "dataset=" + BIOMART_SPECIES_NAME_PLACEHOLDER + "_gene_ensembl",
+            "interface=default",
+            "attributePage=feature_page",
+            "attributeGroup=external",
+            "attributeCollection=microarray"
+    );
+    private static final String UNIPROT_SUFFIX = "_uniprot";
+    private static final String MICROARRAY_SUFFIX = "_microarray";
+    private static final String UNIPROT_SWISSPROT_BIOMART_SEARCH_TERM = "uniprotswissprot";
+    private static final String UNIPROT_TREMBL_BIOMART_SEARCH_TERM = "uniprotsptrembl";
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws IOException - Thrown when file can't be found when during writing and by HTTPConnection class.
```

Comment:
Could "when" be removed here?  Also, changing "and" to "or"?

Reply:
Yes it can. At first I only saw the `Thrown when file can't be found` and I thought this was crazy talk. Not so!